pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,271 @@
+	glog.V(4).Infof("Add job %s to cleanup", job.Name)
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
+		return
+	}
+
+	tc.jobsToCleanup.Add(key)
+}
+
+func (tc *TTLAfterFinishedController) addJobToCleanupAfter(job *batch.Job, after time.Duration) {
+	glog.V(4).Infof("Add job %s to cleanup after %s", job.Name, after.String())
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
+		return
+	}
+
+	tc.jobsToCleanup.AddAfter(key, after)
+}
+
+func (tc *TTLAfterFinishedController) runJobCleanupWorker() {
+	for tc.jobCleanupWorker() {
+	}
+}
+
+func (tc *TTLAfterFinishedController) jobCleanupWorker() bool {
+	key, quit := tc.jobsToCleanup.Get()
+	if quit {
+		return false
+	}
+	defer tc.jobsToCleanup.Done(key)
+
+	err := tc.cleanupJob(key.(string))
+	if err != nil {
+		glog.V(2).Infof("Error cleaning up Job %v: %v", key, err)
+	}
+
+	return true
+}
+
+func (tc *TTLAfterFinishedController) cleanupJob(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Attempt to clean up Job %s", name)
+	// Ignore the Jobs that are already deleted or being deleted
+	job, err := tc.jLister.Jobs(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	if job.DeletionTimestamp != nil {
+		return nil
+	}
+
+	// Before deleting the Job, do a final sanity check just in case TTL is modified.
+	fresh, err := tc.client.BatchV1().Jobs(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Cascade delete the Job when TTL expires
+	deleteJobFn := func(j *batch.Job) error {
+		policy := metav1.DeletePropagationForeground
+		return tc.client.BatchV1().Jobs(j.Namespace).Delete(j.Name, &metav1.DeleteOptions{PropagationPolicy: &policy})
+	}
+	return tc.checkTTLAndReact(fresh, deleteJobFn)
+}
+
+func (tc *TTLAfterFinishedController) checkTTLAndReact(job *batch.Job, runWhenExpired func(job *batch.Job) error) error {
+	if !needsCleanup(job) {
+		return nil
+	}
+
+	now := tc.clock.Now()
+	t, err := timeLeft(job, &now)
+	if err != nil {
+		return err
+	}
+	// If TTL expires, execute runWhenExpired function
+	if *t <= 0 {
+		return runWhenExpired(job)
+	}
+	// If TTL hasn't expired, re-enqueue the Job after t duration
+	tc.addJobToCleanupAfter(job, *t)
+	return nil
+}
+
+func needsCleanup(j *batch.Job) bool {
+	return jobutil.IsJobFinished(j) && j.Spec.TTLSecondsAfterFinished != nil
+}
+
+func timeLeft(j *batch.Job, until *time.Time) (*time.Duration, error) {
+	if !needsCleanup(j) {
+		return nil, fmt.Errorf("Job %s/%s should not be cleaned up", j.Namespace, j.Name)
+	}
+
+	untilUTC := until.UTC()
+	finishAt, err := jobFinishTime(j)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: handle finish time that happens in the future
```

Comment:
You should come up with a well defined behavior for when this occurs. Consider simply deferring destroying the victim after the finish time. It's conservative, but it can't do harm.

Reply:
Done. Because it happens in the future it'll only delay destorying the Job.